File: frontend/src/types/project.ts
import type { VideoCapabilities } from "@/types/auth";

export type SceneTaskStatus = "idle" | "generating" | "success" | "error";

/** One shot inside an episode. Order numbers restart at 1 in every episode. */
export interface Scene {
  id: string;
  episodeId: string | null;
  order: number;
  narration: string;
  /** Spoken by `speakerCharacterId`; `narration` stays the narrator's line. */
  dialogue: string;
  speakerCharacterId: string | null;
  visualPrompt: string;
  shotType: string;
  /** 运镜手法 — how the camera moves through this shot. Reaches the video model. */
  cameraMove: string;
  /** 场景过渡 — how this shot enters from the one before it. */
  transition: string;
  /** The motion prompt. `visualPrompt` describes the frame; this describes the seconds. */
  videoPrompt: string;
  imageReferences: GenerationReferenceInput[];
  videoReferences: GenerationReferenceInput[];
  /** Estimated screen time in ms. 0 means undecided — the project default applies. */
  durationMs: number;
  subtitleText: string;
  /** Who appears in the shot; drives prompt assembly and reference portraits. */
  characterIds: string[];
  /** An approved shot that a batch rerun must leave alone. */
  isLocked: boolean;
  /**
   * Changes only when the server actually changed the row. Key list rows on this rather
   * than on an asset URL: signed links are minted fresh on every response, so a URL-keyed
   * row remounts on every poll and re-downloads its image.
   */
  updatedAt: string;
  image: {
    url: string | null;
    status: SceneTaskStatus;
    progress: number;
  };
  audio: {
    url: string | null;
    status: SceneTaskStatus;
    progress: number;
    duration: number;
  };
  video: {
    url: string | null;
    status: SceneTaskStatus | "idle";
    progress: number;
  };
  /** Last failure reason, persisted so it survives a reload. Empty when the scene is healthy. */
  errorMessage: string;
}

export type EpisodeStatus = "draft" | "storyboard" | "generating" | "done" | "partial" | "failed";

/** An episode without its script or shots: enough to render a switcher or a series list. */
export interface EpisodeSummary {
  id: string;
  projectId: string;
  episodeNumber: number;
  title: string;
  synopsis: string;
  status: EpisodeStatus;
  videoStatus: SceneTaskStatus | "idle";
  videoProgress: number;
  durationMs: number;
  sceneCount: number;
  /** Whether this episode has been anchored yet; the shots are rendered against it. */
  toneImageStatus: SceneTaskStatus | "idle";
  toneImageUrl: string | null;
  errorMessage: string;
  updatedAt: string;
}

export interface Episode extends EpisodeSummary {
  sourceText: string;
  /**
   * The style anchor: a thumbnail grid of every shot, generated in one pass before the
   * full-resolution renders and carried by each of them as a reference. Never a deliverable.
   */
  toneImageUrl: string | null;
  videoUrl: string | null;
  scenes: Scene[];
}

export type ProjectStatus =
  | "idle"
  | "parsing"
  | "generating"
  | "video_generating"
  | "done"
  /** Some shots landed and some failed; retry the failed ones rather than rerunning everything. */
  | "partial"
  | "failed";
export type ProjectMode = "comic" | "drama";
export type ProjectStage = "script" | "bible" | "storyboard" | "audio" | "timeline" | "export";

export interface ProductionSettings {
  mode: ProjectMode;
  aspectRatio: "9:16" | "16:9" | "1:1";
  width: number;
  height: number;
  fps: 24 | 30;
  targetDurationMs: number;
  language: string;
  stylePrompt: string;
  negativePrompt: string;
}

export type ImageResolution = "1K" | "2K" | "4K";
export type GenerationRatio =
  | "auto"
  | "21:9"
  | "16:9"
  | "4:3"
  | "3:2"
  | "1:1"
  | "2:3"
  | "3:4"
  | "9:16"
  | "9:21"
  | "adaptive";
export type VideoQuality = "480p" | "720p" | "1080p" | "2K" | "4K";

/**
 * Which model this series uses for each kind of work, and the parameters every render in
 * it starts from. A null config id means the purpose follows the account default.
 */
export interface ProjectModelSettings {
  textConfigId: number | null;
  imageConfigId: number | null;
  videoConfigId: number | null;
  audioConfigId: number | null;
  imageResolution: ImageResolution;
  imageRatio: GenerationRatio;
  videoQuality: VideoQuality;
  videoAspectRatio: GenerationRatio;
  videoDuration: number;
  videoFps: number;
  videoPromptExtend: boolean;
}

export interface Project {
  id: string;
  title: string;
  /** Shown on the series card. Optional — an empty synopsis renders a placeholder. */
  description: string;
  /** What the user wants the cover to show, in their own words. Drives AI cover generation. */
  coverPrompt: string;
  /** Signed and short-lived, minted per response. Null means "use the fallback cover". */
  coverImageUrl: string | null;
  originalScript: string;
  /** World, tone, and running plot threads, carried into later episodes. */
  seriesBible: string;
  /** The whole cast and every prop, each tiled into one sheet the renderer carries. */
  characterSheetUrl: string | null;
  propSheetUrl: string | null;
  /** Every voice introducing itself, concatenated; a timbre reference for the video model. */
  voiceSheetUrl: string | null;
  status: ProjectStatus;
  videoStatus: SceneTaskStatus | "idle";
  videoProgress: number;
  videoUrl: string | null;
  productionSettings: ProductionSettings;
  modelSettings: ProjectModelSettings;
  currentStage: ProjectStage;
  /** The episode `scenes` belongs to. Null only for a series with no episode yet. */
  currentEpisodeId: string | null;
  episodes: EpisodeSummary[];
  updatedAt: string;
  /** The current episode's shots, never the whole series': order numbers restart each episode. */
  scenes: Scene[];
}

export interface ProjectItemResponse {
  project: Project;
}

export interface ProjectListResponse {
  projects: Project[];
}

export interface CreateProjectInput {
  title?: string;
  description?: string;
  coverPrompt?: string;
  originalScript?: string;
  productionSettings?: Partial<ProductionSettings>;
}

export interface UpdateProjectInput {
  title?: string;
  description?: string;
  coverPrompt?: string;
  originalScript?: string;
  seriesBible?: string;
  /**
   * Send only what changed. A config id of `0` clears the pick and returns that purpose to
   * the account default — `null` cannot mean "clear", because in a PATCH it already means
   * "leave alone".
   */
  modelSettings?: Partial<ProjectModelSettings>;
}

export interface SetProjectCoverInput {
  /** A `data:image/png|jpeg|webp;base64,` string. The API never takes multipart. */
  imageData: string;
}

/**
 * Cover generation is project-less on purpose: the create dialog runs it before a project
 * exists, and the edit dialog runs it on unsaved edits. It never writes — the caller
 * previews the result and applies it with `setProjectCoverAction`.
 */
export interface GenerateCoverInput {
  /** What the cover should show. Required — this is the subject of the picture. */
  prompt: string;
  title?: string;
  stylePrompt?: string;
}

export interface GenerateCoverResponse {
  imageData: string;
}

/** One resolved model, with the limits the UI must not let the user exceed. */
export interface ProjectModelSummary {
  provider: string;
  model: string;
  source: string;
  configId: number | null;
  /** False when this purpose is falling through to the account default. */
  isProjectPick: boolean;
  capabilities?: {
    maxReferenceImages?: number;
    resolutions?: ImageResolution[];
    ratios?: GenerationRatio[];
  } | null;
}

export interface ProjectModelsResponse {
  models: {
    text: ProjectModelSummary | null;
    image: ProjectModelSummary | null;
    video: (ProjectModelSummary & { capabilities: VideoCapabilities | null }) | null;
    audio: ProjectModelSummary | null;
  };
  modelSettings: ProjectModelSettings;
}

export interface PromptPreset {
  key: string;
  label: string;
  template: string;
}

export interface PromptPresetListResponse {
  kind: string;
  presets: PromptPreset[];
}

export interface CreateEpisodeInput {
  /** Required: a list of "第 1 集 / 第 2 集" tells the user nothing. */
  title: string;
  synopsis?: string;
  sourceText?: string;
}

/** Which half of the storyboard a breakdown produces. */
export type BreakdownTarget = "shots" | "video" | "both";

/**
 * What the breakdown may look at. Selecting nothing is meaningful: it tells the model to
 * decide everything from the script alone. Characters the bible has never heard of are
 * inferred from the script either way.
 */
export interface BreakdownReferences {
  characterIds?: string[];
  propIds?: string[];
  voiceProfileIds?: string[];
  useCastSheet?: boolean;
  usePropSheet?: boolean;
  useVoiceSheet?: boolean;
}

export interface BreakdownEpisodeInput {
  target: BreakdownTarget;
  script?: string;
  references?: BreakdownReferences;
  /** Re-splitting discards rendered shots; the first call reports, this confirms. */
  replaceAll?: boolean;
  model?: string;
}

export interface BreakdownEpisodeResponse {
  projectId: string;
  episodeId: string;
  target: BreakdownTarget;
  applied: boolean;
  discardsGeneratedScenes: number;
  shotCount?: number;
  scenes: Scene[];
}

/** Anchor an episode's look without rendering any shots against it yet. */
export interface GenerateToneSheetInput {
  previousEpisodeId?: string;
  mergeReferences?: boolean;
  regenerate?: boolean;
  references?: GenerationReferenceInput[];
}

export interface GenerateToneSheetResponse {
  projectId: string;
  episodeId: string;
  status: ProjectStatus;
  regeneratesToneSheet: boolean;
}

export interface CancelProjectRunResponse {
  projectId: string;
  canceled: boolean;
  status: ProjectStatus;
}

/**
 * Render an episode: one tone sheet to anchor the look, then a frame per shot.
 *
 * `mergeReferences` trades tokens for fidelity — merged, the cast sheet, prop sheet, and
 * previous anchor arrive as one image; separate, they cost more. `regenerate` resamples the
 * anchor instead of reusing it, which also restyles shots that were already approved.
 */
export interface GenerateStoryboardInput {
  previousEpisodeId?: string;
  mergeReferences?: boolean;
  regenerate?: boolean;
  sceneIds?: string[];
  references?: GenerationReferenceInput[];
  /**
   * Render only the shots that are not already rendered. What "retry the failed ones" means:
   * without it a rerun re-renders, and re-pays for, every unlocked shot in the episode.
   */
  pendingOnly?: boolean;
}

export type GenerationReferenceKind = "character" | "characterState" | "prop" | "tone" | "sceneImage" | "sceneVideo" | "voice";

export interface GenerationReferenceInput {
  kind: GenerationReferenceKind;
  id: string;
}

export interface GenerateStoryboardResponse {
  projectId: string;
  episodeId: string;
  status: ProjectStatus;
  shotCount: number;
  referenceCount: number;
  mergeReferences: boolean;
  regeneratesToneSheet: boolean;
}

export interface UpdateEpisodeInput {
  title?: string;
  synopsis?: string;
  sourceText?: string;
  status?: EpisodeStatus;
}

export interface EpisodeListResponse {
  episodes: EpisodeSummary[];
}

export interface EpisodeItemResponse {
  episode: Episode;
}

/**
 * One look a character can appear in: an age, an outfit, a transformation. States are
 * parallel forms of the same person, not a timeline — an episode range is an optional
 * narrowing, and a state that leaves it unset is simply always available.
 */
export interface CharacterState {
  id: string;
  characterId: string;
  name: string;
  /** What this state is ("十六岁，校服"); feeds the model that drafts `finalPrompt`. */
  description: string;
  /** Empty means "the look did not change"; the card's own prompt still applies. */
  appearancePrompt: string;
  /** The reviewed prompt that actually draws the sheet. Shown in the UI simply as 提示词. */
  finalPrompt: string;
  /** The turnaround sheet: front, three-quarter, and profile in one image. */
  referenceImageUrl: string | null;
  voiceModel: string;
  orderNum: number;
  /** Null on either means the state is not pinned to an episode range at all. */
  fromEpisode: number | null;
  toEpisode: number | null;
  updatedAt: string;
}

/** A recurring cast member, pinned to one look, one image model, and one voice. */
export interface Character {
  id: string;
  projectId: string;
  name: string;
  /** Comma-separated names the script also uses for this character. */
  aliases: string;
  description: string;
  appearancePrompt: string;
  /** Legacy card portrait from before looks were split into states; nothing writes it. */
  referenceImageUrl: string | null;
  /** Every state of this character tiled into one image. */
  sheetImageUrl: string | null;
  /** Frozen when the first image was drawn, so a later default change cannot restyle them. */
  imageProvider: string;
  imageModel: string;
  voiceProvider: string;
  voiceModel: string;
  /** The bound voice profile, if any; it wins over the two columns above. */
  voiceProfileId: string | null;
  /** Set once the look is approved; a locked card is never redrawn. */
  isLocked: boolean;
  orderNum: number;
  states: CharacterState[];
  updatedAt: string;
}

export interface CreateCharacterInput {
  name: string;
  aliases?: string;
  description?: string;
  appearancePrompt?: string;
  voiceProvider?: string;
  voiceModel?: string;
  orderNum?: number;
}

export interface UpdateCharacterInput {
  name?: string;
  aliases?: string;
  description?: string;
  appearancePrompt?: string;
  voiceProvider?: string;
  voiceModel?: string;
  /** "" unbinds. Not null: the backend reads an absent field as "leave it alone". */
  voiceProfileId?: string;
  isLocked?: boolean;
  orderNum?: number;
}

export interface CreateCharacterStateInput {
  name: string;
  description?: string;
  appearancePrompt?: string;
  finalPrompt?: string;
  voiceModel?: string;
  orderNum?: number;
  fromEpisode?: number | null;
  toEpisode?: number | null;
}

export type UpdateCharacterStateInput = Partial<CreateCharacterStateInput>;

/**
 * Draft an image prompt for review. The fields come from the dialog rather than the stored
 * row so drafting works against edits the user has not saved yet. The instruction template
 * is the built-in one and is not overridable; `preset` picks which built-in to draft from.
 */
export interface DraftPromptInput {
  name?: string;
  description?: string;
  preset?: string;
  model?: string;
}

export interface DraftPromptResponse {
  prompt: string;
}

/** Empty falls back to the stored `finalPrompt`. */
export interface GenerateReferenceImageInput {
  prompt?: string;
}

export interface UploadReferenceImageInput {
  /** A `data:image/png|jpeg|webp;base64,` string. The API never takes multipart. */
  imageData: string;
}

export interface CharacterListResponse {
  characters: Character[];
}

export interface CharacterItemResponse {
  character: Character;
}

export interface CharacterStateItemResponse {
  state: CharacterState;
}

export interface CastSheetResponse {
  characterSheetUrl: string | null;
}

/** One voice in the show: the narrator, or a character's timbre. */
export interface VoiceProfile {
  id: string;
  projectId: string;
  name: string;
  /** Free-text note for the user ("沙哑，压低"); never sent to a provider. */
  note: string;
  /** A provider and a model, never credentials — those live on the account's audio config. */
  voiceProvider: string;
  voiceModel: string;
  /** The line this voice says in the merged track; the wording tells the model when to use it. */
  sampleText: string;
  audioUrl: string | null;
  orderNum: number;
  updatedAt: string;
}

export interface CreateVoiceProfileInput {
  name: string;
  note?: string;
  voiceProvider?: string;
  voiceModel?: string;
  sampleText?: string;
  orderNum?: number;
}

export type UpdateVoiceProfileInput = Partial<CreateVoiceProfileInput>;

/**
 * Design a timbre from a description and bind it to this series in one step. Provider and
 * model come from the project's audio configuration, never from the client.
 */
export interface DesignVoiceProfileInput {
  name: string;
  voicePrompt: string;
  /** What the audition says. Distinct from `sampleText`, the line in the merged track. */
  previewText: string;
  note?: string;
  sampleText?: string;
}

/** Bind a timbre already saved on the account to this series. */
export interface ImportVoiceProfileInput {
  userVoiceId: string;
  name?: string;
  note?: string;
  sampleText?: string;
}

export interface VoiceProfileListResponse {
  voices: VoiceProfile[];
}

export interface VoiceProfileItemResponse {
  voice: VoiceProfile;
}

export interface VoiceSheetResponse {
  voiceSheetUrl: string | null;
}

/** An object the series has to draw the same way every time it appears. */
export interface Prop {
  id: string;
  projectId: string;
  name: string;
  description: string;
  /** Whose prop it is; printed on the reference image. Null leaves it unattributed. */
  ownerCharacterId: string | null;
  /** Resolved by the server so a card can render the owner without a second request. */
  ownerName: string;
  /** The reviewed prompt that actually draws the image. Shown in the UI simply as 提示词. */
  finalPrompt: string;
  imageUrl: string | null;
  orderNum: number;
  updatedAt: string;
}

export interface CreatePropInput {
  name: string;
  description?: string;
  /** "" leaves it unattributed. Not null: an absent field means "leave alone" in a PATCH. */
  ownerCharacterId?: string;
  finalPrompt?: string;
  orderNum?: number;
}

export type UpdatePropInput = Partial<CreatePropInput>;

export interface PropListResponse {
  props: Prop[];
}

export interface PropItemResponse {
  prop: Prop;
}

export interface PropSheetResponse {
  propSheetUrl: string | null;
}

export interface SetSceneCastInput {
  characterIds: string[];
}

export interface SetSceneCastResponse {
  sceneId: string;
  characterIds: string[];
}

export type UpdateProductionSettingsInput = Partial<ProductionSettings> & {
  currentStage?: ProjectStage;
};

export type GenerationJobStatus = "queued" | "running" | "succeeded" | "failed" | "canceled";

export interface GenerationJob {
  id: string;
  projectId: string;
  sceneId: string | null;
  jobType: "storyboards" | "audio" | "videos" | "preview" | "export";
  status: GenerationJobStatus;
  progress: number;
  attempt: number;
  maxAttempts: number;
  errorCode: string | null;
  errorMessage: string | null;
  createdAt: string;
  updatedAt: string;
  startedAt: string | null;
  finishedAt: string | null;
}

export interface GenerationJobListResponse {
  jobs: GenerationJob[];
}

export interface UpdateSceneInput {
  narration?: string;
  dialogue?: string;
  speakerCharacterId?: string;
  visualPrompt?: string;
  shotType?: string;
  cameraMove?: string;
  transition?: string;
  videoPrompt?: string;
  imageReferences?: GenerationReferenceInput[];
  videoReferences?: GenerationReferenceInput[];
  durationMs?: number;
  subtitleText?: string;
  isLocked?: boolean;
}
export interface CreateSceneInput extends UpdateSceneInput {
  episodeId?: string;
}
export interface ReorderScenesInput {
  sceneIds: string[];
  /** Order numbers restart each episode, so a reorder is always within one. */
  episodeId?: string;
}

export interface ParseProjectInput {
  script: string;
  model?: string;
  /** Which episode the parsed shots land in. Omitted means the current one. */
  episodeId?: string;
  /**
   * Reparsing discards generated shots. Without this the backend returns a preview
   * (`applied: false`) so the user can confirm before losing rendered work.
   */
  replaceAll?: boolean;
}

/** A parsed shot that has not been written to the project yet. */
export interface PendingScene {
  order: number;
  narration: string;
  visualPrompt: string;
}

export interface ParseProjectResponse {
  projectId: string;
  episodeId: string;
  status: ProjectStatus;
  source: "llm" | "fallback";
  warning?: string;
  applied: boolean;
  discardsGeneratedScenes: number;
  pendingScenes: PendingScene[];
  scenes: Scene[];
}

export interface GenerateProjectInput {
  model?: string;
  episodeId?: string;
  sceneIds?: string[];
  /** See `GenerateStoryboardInput.pendingOnly`. */
  pendingOnly?: boolean;
}

export interface GenerateProjectResponse {
  projectId: string;
  episodeId: string;
  status: ProjectStatus;
  sceneCount: number;
  model?: string;
  provider?: string;
  imageModel?: string;
}

export interface OptimizeProjectInput {
  script?: string;
  model?: string;
}

export interface OptimizeProjectResponse {
  projectId: string;
  optimizedScript: string;
  tips: string[];
  source: "llm" | "fallback";
  warning?: string;
  appliedToProject: boolean;
}

export interface GenerateVideoInput {
  model?: string;
  episodeId?: string;
  quality?: string;
  aspectRatio?: string;
  fps?: number;
  duration?: number;
  promptExtend?: boolean;
  /**
   * Pass the project's merged timbre reference. Costs more, and a model that does not accept
   * reference audio returns 400 rather than silently dropping it.
   */
  withAudio?: boolean;
  sceneIds?: string[];
  references?: GenerationReferenceInput[];
  /** See `GenerateStoryboardInput.pendingOnly`. A clip is the costliest thing to redo. */
  pendingOnly?: boolean;
}

export interface GenerateVideoResponse {
  projectId: string;
  status: ProjectStatus;
  model: string;
  episodeId: string;
  sceneCount: number;
  withAudio: boolean;
}

export type ExportStatus = "queued" | "running" | "succeeded" | "failed" | "canceled";

/** Several rendered shots merged into one file. Order follows the request, not shot number. */
export interface ExportJob {
  id: string;
  projectId: string;
  sceneIds: string[];
  rangeLabel: string;
  status: ExportStatus;
  progress: number;
  videoUrl: string | null;
  fileSize: number;
  errorMessage: string;
  createdAt: string;
  updatedAt: string;
  finishedAt: string | null;
}

export interface CreateExportInput {
  /** Ordered: the video section assembles a cut, which need not follow the storyboard. */
  sceneIds: string[];
  rangeLabel?: string;
}

export interface ExportListResponse {
  exports: ExportJob[];
}

export interface ExportItemResponse {
  export: ExportJob;
}

export interface SceneUpdatePayload {
  episodeId?: string;
  narration?: string;
  dialogue?: string;
  speakerCharacterId?: string | null;
  characterIds?: string[];
  visualPrompt?: string;
  shotType?: string;
  cameraMove?: string;
  durationMs?: number;
  subtitleText?: string;
  isLocked?: boolean;
  order?: number;
  parseStatus?: string;
  imageStatus?: SceneTaskStatus;
  imageProgress?: number;
  imageUrl?: string | null;
  audioStatus?: SceneTaskStatus;
  audioProgress?: number;
  audioUrl?: string | null;
  audioDuration?: number;
  videoStatus?: SceneTaskStatus | "idle";
  videoProgress?: number;
  videoUrl?: string | null;
  videoModel?: string;
  errorMsg?: string;
}

File: backend/app/models/project.py
from __future__ import annotations

from sqlalchemy import CheckConstraint, Index, text
from sqlmodel import Field, SQLModel


class Project(SQLModel, table=True):
    """A series. Its content lives in `Episode` rows, not directly on the project."""

    __tablename__ = "projects"
    __table_args__ = (
        CheckConstraint("mode IN ('comic', 'drama')"),
        Index("idx_projects_user_id", "user_id"),
        Index("idx_projects_deleted_at", "deleted_at"),
    )

    id: str = Field(primary_key=True)
    created_at: str | None = None
    updated_at: str | None = None
    deleted_at: str | None = None
    user_id: int
    title: str | None = None
    # Shown on the series card. Optional: a project with no synopsis falls back to a
    # placeholder in the UI rather than blocking creation.
    description: str = Field(default="", sa_column_kwargs={"server_default": text("''")})
    # A path relative to PRIVATE_GENERATED_DIR, never a URL — same rule as every other
    # asset column here, because signed links expire after ARTIFACT_TTL_DAYS.
    cover_image_path: str | None = None
    original_script: str | None = None
    status: str | None = Field(default="idle", sa_column_kwargs={"server_default": text("'idle'")})
    video_url: str | None = None
    video_status: str | None = Field(default="idle", sa_column_kwargs={"server_default": text("'idle'")})
    video_progress: int | None = Field(default=0, sa_column_kwargs={"server_default": text("0")})
    mode: str = Field(default="comic", sa_column_kwargs={"server_default": text("'comic'")})
    aspect_ratio: str = Field(default="9:16", sa_column_kwargs={"server_default": text("'9:16'")})
    width: int = Field(default=1080, sa_column_kwargs={"server_default": text("1080")})
    height: int = Field(default=1920, sa_column_kwargs={"server_default": text("1920")})
    fps: int = Field(default=24, sa_column_kwargs={"server_default": text("24")})
    target_duration_ms: int = Field(default=60000, sa_column_kwargs={"server_default": text("60000")})
    language: str = Field(default="zh-CN", sa_column_kwargs={"server_default": text("'zh-CN'")})
    style_prompt: str = Field(default="", sa_column_kwargs={"server_default": text("''")})
    negative_prompt: str = Field(default="", sa_column_kwargs={"server_default": text("''")})
    current_stage: str = Field(default="script", sa_column_kwargs={"server_default": text("'script'")})
    # World, tone, and running plot threads. Fed to the model when drafting the next episode
    # so a long-running series stays coherent instead of restarting from scratch each time.
    series_bible: str = Field(default="", sa_column_kwargs={"server_default": text("''")})
    # The whole cast, and every prop, each tiled into one sheet. Rebuilt on demand and
    # carried into storyboard renders: providers cap how many reference images one request
    # may hold, so a series of any size has to arrive as a couple of sheets, not a crowd.
    character_sheet_path: str | None = None
    prop_sheet_path: str | None = None
    # Every voice in the show introducing itself in its own timbre, concatenated into one
    # track. Passed to the video model as a reference so it can keep speakers apart.
    voice_sheet_path: str | None = None
    # What the user wants the cover to show, in their own words. The cover used to be drawn
    # from title + synopsis, which describes the story rather than the picture — a series
    # about a betrayal got a picture of a betrayal, never the poster the user had in mind.
    cover_prompt: str = Field(default="", sa_column_kwargs={"server_default": text("''")})
    # The model configuration this series uses for each kind of work. Resolution is
    # project-first: an unset column falls back to the account's active config, so existing
    # projects keep working and a user who never opens the model panel notices nothing.
    #
    # Deliberately not foreign keys, for the same reason as `characters.voice_profile_id`:
    # SQLite cannot add a constrained column in place, and a batch recreate of this table
    # reflects into the shared SQLModel metadata and reorders columns unpredictably. A
    # deleted config simply stops resolving and the fallback takes over.
    text_config_id: int | None = None
    image_config_id: int | None = None
    video_config_id: int | None = None
    audio_config_id: int | None = None
    # Generation defaults carried into every render this series starts. They live here
    # rather than being re-picked per run so a storyboard and the clips made from it agree,
    # and so the episode editor can prefill without asking the same six questions each time.
    # Validated against the selected model's declared capabilities on write.
    image_resolution: str = Field(default="2K", sa_column_kwargs={"server_default": text("'2K'")})
    image_ratio: str = Field(default="auto", sa_column_kwargs={"server_default": text("'auto'")})
    video_quality: str = Field(default="720p", sa_column_kwargs={"server_default": text("'720p'")})
    # Separate from `aspect_ratio` above: that one is the finished canvas, this is the
    # parameter handed to the video model, and the two provider vocabularies differ.
    video_aspect_ratio: str = Field(default="9:16", sa_column_kwargs={"server_default": text("'9:16'")})
    video_duration: int = Field(default=5, sa_column_kwargs={"server_default": text("5")})
    video_fps: int = Field(default=24, sa_column_kwargs={"server_default": text("24")})
    video_prompt_extend: bool = Field(default=False, sa_column_kwargs={"server_default": text("0")})


class Episode(SQLModel, table=True):
    """One installment of a series: independently written, storyboarded, and rendered."""

    __tablename__ = "episodes"
    __table_args__ = (
        CheckConstraint("episode_number > 0"),
        Index(
            "idx_episodes_project_number",
            "project_id",
            "episode_number",
            unique=True,
            sqlite_where=text("deleted_at IS NULL"),
        ),
        Index("idx_episodes_deleted_at", "deleted_at"),
    )

    id: str = Field(primary_key=True)
    created_at: str | None = None
    updated_at: str | None = None
    deleted_at: str | None = None
    project_id: str = Field(foreign_key="projects.id", ondelete="CASCADE")
    episode_number: int
    title: str = Field(default="", sa_column_kwargs={"server_default": text("''")})
    # Short recap carried into later episodes as context; cheaper than replaying full scripts.
    synopsis: str = Field(default="", sa_column_kwargs={"server_default": text("''")})
    source_text: str = Field(default="", sa_column_kwargs={"server_default": text("''")})
    status: str = Field(default="draft", sa_column_kwargs={"server_default": text("'draft'")})
    # A thumbnail grid of every shot in this episode, generated in one pass before the
    # full-resolution renders. It exists to be a *style anchor*, not a deliverable: one
    # sampling fixes lighting, colour, and render style for the whole episode, and each
    # per-shot render then carries it as a reference. Without it the shots agree on faces
    # (via the cast sheet) and on nothing else.
    tone_image_path: str | None = None
    tone_image_status: str = Field(default="idle", sa_column_kwargs={"server_default": text("'idle'")})
    video_path: str | None = None
    video_status: str = Field(default="idle", sa_column_kwargs={"server_default": text("'idle'")})
    video_progress: int = Field(default=0, sa_column_kwargs={"server_default": text("0")})
    duration_ms: int = Field(default=0, sa_column_kwargs={"server_default": text("0")})
    error_message: str | None = None


class Scene(SQLModel, table=True):
    """One shot inside an episode."""

    __tablename__ = "scenes"
    __table_args__ = (
        Index("idx_scenes_project_id", "project_id"),
        Index("idx_scenes_episode_order", "episode_id", "order_num"),
        Index("idx_scenes_deleted_at", "deleted_at"),
    )

    id: str = Field(primary_key=True)
    created_at: str | None = None
    updated_at: str | None = None
    deleted_at: str | None = None
    project_id: str = Field(foreign_key="projects.id", ondelete="CASCADE")
    episode_id: str | None = Field(default=None, foreign_key="episodes.id", ondelete="CASCADE")
    order_num: int | None = None
    narration: str | None = None
    # Spoken by `speaker_character_id`, so it is voiced with that character's locked voice.
    # `narration` stays the narrator's line.
    dialogue: str = Field(default="", sa_column_kwargs={"server_default": text("''")})
    speaker_character_id: str | None = None
    visual_prompt: str | None = None
    shot_type: str = Field(default="", sa_column_kwargs={"server_default": text("''")})
    # 运镜手法 — how the camera moves through this shot. Written by the storyboard breakdown
    # and editable afterwards; it reaches the video model, not the still renderer.
    camera_move: str = Field(default="", sa_column_kwargs={"server_default": text("''")})
    # 场景过渡 — how this shot enters from the one before it (hard cut, dissolve, fade).
    # A property of the seam rather than the frame, which is why it never reaches the
    # prompt that draws the still.
    transition: str = Field(default="", sa_column_kwargs={"server_default": text("''")})
    # The motion prompt, kept apart from `visual_prompt`. One describes a frame, the other
    # describes what happens over several seconds; collapsing them produced clips that
    # either stood still or ignored the composition the storyboard image had already fixed.
    video_prompt: str = Field(default="", sa_column_kwargs={"server_default": text("''")})
    # The shot's screen time. Written by the breakdown as an estimate and editable after;
    # 0 means "undecided", and the renderer falls back to the project default.
    duration_ms: int = Field(default=0, sa_column_kwargs={"server_default": text("0")})
    subtitle_text: str = Field(default="", sa_column_kwargs={"server_default": text("''")})
    # Asset columns hold a path relative to PRIVATE_GENERATED_DIR, never a signed URL:
    # links expire after ARTIFACT_TTL_DAYS and are minted per response by the serializer.
    image_path: str | None = None
    image_status: str | None = Field(default="idle", sa_column_kwargs={"server_default": text("'idle'")})
    audio_path: str | None = None
    audio_status: str | None = Field(default="idle", sa_column_kwargs={"server_default": text("'idle'")})
    audio_duration: float = Field(default=0, sa_column_kwargs={"server_default": text("0")})
    # Only used in drama mode, where a shot is upgraded from a still to a generated clip.
    video_path: str | None = None
    video_status: str = Field(default="idle", sa_column_kwargs={"server_default": text("'idle'")})
    # An approved shot a batch rerun must leave alone.
    is_locked: bool = Field(default=False, sa_column_kwargs={"server_default": text("0")})
    error_message: str | None = None


class GenerationJob(SQLModel, table=True):
    __tablename__ = "generation_jobs"
    __table_args__ = (
        CheckConstraint("status IN ('queued', 'running', 'succeeded', 'failed', 'canceled')"),
        Index("idx_generation_jobs_project_created", "project_id", text("created_at DESC")),
        Index("idx_generation_jobs_status_lease", "status", "lease_expires_at", "created_at"),
        Index(
            "idx_generation_jobs_idempotency",
            "user_id",
            "project_id",
            "idempotency_key",
            unique=True,
            sqlite_where=text("idempotency_key IS NOT NULL"),
        ),
    )

    id: str = Field(primary_key=True)
    created_at: str
    updated_at: str
    started_at: str | None = None
    finished_at: str | None = None
    user_id: int = Field(foreign_key="users.id", ondelete="CASCADE")
    project_id: str = Field(foreign_key="projects.id", ondelete="CASCADE")
    episode_id: str | None = Field(default=None, foreign_key="episodes.id", ondelete="CASCADE")
    scene_id: str | None = Field(default=None, foreign_key="scenes.id", ondelete="CASCADE")
    job_type: str
    status: str = Field(default="queued", sa_column_kwargs={"server_default": text("'queued'")})
    progress: int = Field(default=0, sa_column_kwargs={"server_default": text("0")})
    input_json: str = Field(default="{}", sa_column_kwargs={"server_default": text("'{}'")})
    result_json: str | None = None
    attempt: int = Field(default=0, sa_column_kwargs={"server_default": text("0")})
    max_attempts: int = Field(default=3, sa_column_kwargs={"server_default": text("3")})
    idempotency_key: str | None = None
    lease_owner: str | None = None
    lease_expires_at: str | None = None
    heartbeat_at: str | None = None
    error_code: str | None = None
    error_message: str | None = None
